### ISSUE TITLE
Option to have batches only contain messages of the same key, where no other batch inflight would contain said key

### ISSUE THREAD

**Issue by astubbs** [confluentinc/parallel-consumer#266]
Imagine a scenario where you have a topic where key is the account. And you have databases per account. You want to do batch inserts to the database, where you do a batch insert per message from within the batch. 

You wouldn't want any messages from a different key, in this batch.

**Comment by timcrouzet** [comment]
Hi @astubbs 

Thanks very much for the great work in this library!

I have a scenario that is exactly like the one you describe and I was thinking of using parallel-consumer to implement it. 

My scenario : 

- I have a topic where the key is the date (YYYYMMDD)
- I have one database table per date
- I want to do batch inserts by database table (one insert per message). So I need to run a SQL batch with messages having the same key. 
- Ordering is required.


My understanding (But it looks like I am wrong) is that when adding a batchSize, my user function (executing the SQL inserts) will be called with a list of records of the same key.
So the parallel consumer shall already be distributing batches of messages of the same key to the different threads running user function?

But your message suggests that this is not the case. What did I get wrong here ?

Also, do you have an idea of a potential release date for 0.5.0.0 where this feature will be available ?

Thanks very much

Tim

**Comment by astubbs** [comment]
Yes atm this doesn't happen. You /can/ get this with partition or unordered mode if you're lucky, but it's not guaranteed. If using KEY ordering mode, it's guaranteed NOT to happen currently. 

0.5 is about to be released as is (won't have this feature). 

I'll need to have a look to see how easy this will be to implement before giving an estimate - also depends on demand. But I think this is a useful feature to have. 

Thanks for participating! 🤗

**Comment by timcrouzet** [comment]
Thanks for the prompt answer! Excellent news about 0.5, is it a matter of hours or a few days ? 

Just to confirm my understanding about the batching.

If I use the below (Key based ordering + batching) then I shall get 5 threads running my 'processSQLBatch' function with each 'payload' containing only messages of the same key?

```
ParallelStreamProcessor.createEosStreamProcessor(ParallelConsumerOptions.<String, String>builder()
        .consumer(getKafkaConsumer())
        .ordering(KEY)
        .maxConcurrency(5)
        .batchSize(1000)
        .build());

parallelConsumer.poll(context -> {
    List<String> payload = context.stream()
            .map(this::preparePayload)
            .collect(Collectors.toList());
    // process the entire batch payload at once
    processSQLBatch(payload);
});
```

**Comment by astubbs** [comment]
Regarding 0.5 btw, at this point I'm not aware of anything substantial that is going to change from the existing released 0.5 snapshot.

**Comment by astubbs** [comment]
Also regarding the new option that I'm referring to, I would really appreciate your feedback as to what do you think the API should look like.

For example - to represent this new feature, should it be a new ordering mode, or an option that you enable when using the key ordering mode. In which case wouldn't labelling this option while using a mode other than key, cause an error?

**Comment by mbieser** [comment]
+1
This is common use case for my project too.

**Comment by yura-arab4uk** [comment]
+1
This feature is really necessary

**Comment by mageddo** [comment]
In my project I want to reduce the messages before do the operations at the database for example, if my consumer is responsible for do balance operations and in the batch I have 3 messages of credit account operations with the same key 

```
+10
+4
+1
```

Then I can reduce them to `15` and do only one credit operation on the database instead of 3, which improves a lot of performance and reduce concurrence.

**Comment by astubbs** [comment]
Tracked in the maintained fork as astubbs/parallel-consumer#145, which carries a summary of this
issue and its current status there.

The **Confluent-hosted version** of this project is no longer maintained - its README says so.
**Parallel Consumer itself continues** in my fork: https://github.com/astubbs/parallel-consumer

Nothing in this repository is closed or changed by this comment.
